edward/inferences/hmcda.py

```diff
@@ -0,0 +1,331 @@
+    --------
+    >>> z = Normal(loc=0.0, scale=1.0)
+    >>> x = Normal(loc=tf.ones(10) * z, scale=1.0)
+    >>>
+    >>> qz = Empirical(tf.Variable(tf.zeros(500)))
+    >>> data = {x: np.array([0.0] * 10, dtype=np.float32)}
+    >>> inference = ed.HMCDA({z: qz}, data)
+    """
+    super(HMCDA, self).__init__(*args, **kwargs)
+
+  def initialize(self, n_adapt, delta=0.65, Lambda=0.15, *args, **kwargs):
+    """
+    Parameters
+    ----------
+    n_adapt : float
+      Number of samples with adaption for epsilon
+    delta : float, optional
+      Target accept rate
+    Lambda : float, optional
+      Target leapfrog length
+    """
+    self.scope_iter = 0  # a convenient counter for log joint calculations
+
+    # Find intial epsilon
+    step_size = self.find_good_eps()
+    sess = get_session()
+    init_op = tf.global_variables_initializer()
+    sess.run(init_op)
+    step_size = sess.run(step_size)
+
+    # Variables for Dual Averaging
+    self.epsilon = tf.Variable(step_size, trainable=False)
+    self.mu = tf.cast(tf.log(10.0 * step_size), tf.float32)
+    self.epsilon_B = tf.Variable(1.0, trainable=False, name="epsilon_bar")
+    self.H_B = tf.Variable(0.0, trainable=False, name="H_bar")
+
+    # Parameters for Dual Averaging
+    self.n_adapt = n_adapt
+    self.delta = tf.constant(delta, dtype=tf.float32)
+    self.Lambda = tf.constant(Lambda)
+
+    self.gamma = tf.constant(0.05)
+    self.t_0 = tf.constant(10)
+    self.kappa = tf.constant(0.75)
+
+    return super(HMCDA, self).initialize(*args, **kwargs)
+
+  def build_update(self):
+    """Simulate Hamiltonian dynamics using a numerical integrator.
+    Correct for the integrator's discretization error using an
+    acceptance ratio. The initial value of espilon is heuristically chosen
+    with Algorithm 4
+
+    Notes
+    -----
+    The updates assume each Empirical random variable is directly
+    parameterized by ``tf.Variable``s.
+    """
+
+    old_sample = {z: tf.gather(qz.params, tf.maximum(self.t - 1, 0))
+                  for z, qz in six.iteritems(self.latent_vars)}
+    old_sample = OrderedDict(old_sample)
+
+    # Sample momentum.
+    old_r_sample = OrderedDict()
+    for z, qz in six.iteritems(self.latent_vars):
+      event_shape = qz.event_shape
+      normal = Normal(loc=tf.zeros(event_shape), scale=tf.ones(event_shape))
+      old_r_sample[z] = normal.sample()
+
+    # Simulate Hamiltonian dynamics.
+    L_m = tf.maximum(1, tf.cast(tf.round(self.Lambda / self.epsilon), tf.int32))
+    new_sample, new_r_sample = leapfrog(old_sample, old_r_sample,
+                                        self.epsilon, self._log_joint, L_m)
+
+    # Calculate acceptance ratio.
+    ratio = kinetic_energy(old_r_sample)
+    ratio -= kinetic_energy(new_r_sample)
+    ratio += self._log_joint(new_sample)
+    ratio -= self._log_joint(old_sample)
+
+    # Accept or reject sample.
+    u = Uniform().sample()
+    alpha = tf.minimum(1.0, tf.exp(ratio))
+    accept = u < alpha
```

Comment:
Comparing with `tf.log(u) < ratio` should be more numerically stable than checking on the PDF scale.